src/Workspaces/Core/Portable/FlowAnalysis/SymbolUsageAnalysis/SymbolUsageAnalysis.DataFlowAnalyzer.FlowGraphAnalysisData.cs

```diff
@@ -548,8 +561,10 @@ public override bool TryGetDelegateInvocationTargets(IOperation write, out Immut
                     return false;
                 }
 
-                protected override void DisposeCoreData()
+                public override void Dispose()
                 {
+                    base.Dispose();
```

Comment:
oddly enough, my strong preference is to avoid approaches that require overrides to remember to call into their base.  This can be done by isntead having the base call into an abstract that derived must supply.  this means the base always executes its loic, and the derived also knows it must consider how it should be disposed.

Reply:
Tag @sharwell - I personally prefer his recommended approach of virtual/override with the override always expected to call base.Dispose. Sam, do we have a consistent pattern used across Roslyn? If so, I would prefer to stick to that.

Reply:
i'm fine with it as well.  would def like analyzer to catch subclasses forgetting to call base.

Reply:
We have https://docs.microsoft.com/en-us/visualstudio/code-quality/ca2215?view=vs-2019 implemented in FxCop analyzers package. We have disabled it for Roslyn in the ruleset: https://github.com/dotnet/roslyn/blob/342b2712e2db77cdb270eb948616406b6b75550f/eng/config/rulesets/Shipping.ruleset#L62

Reply:
@CyrusNajmabadi The approach you describe only works for a 2-layer hierarchy. The approach used here works equally well for _n_ layers.

Reply:
it def does :)

abstract class B1
{
     public void Dispose() { DerivedDispose1(); /* now my cleanup */ }
     protected abstract void DerivedDispose1();
}

abstract class B2 : B1
{
    protected sealed override DerivedDispose()
    {
        DerivedDispose2();
        /* now my cleanup */
    }
     protected abstract void DerivedDispose2();
}
```

Reply:
@CyrusNajmabadi that keeps adding complexity, and it still only works consistently when every type in the hierarchy is _either_ `abstract` or `sealed`.

Reply:
>  and it still only works consistently when every type in the hierarchy is either abstract or sealed.

I would def hope that's what we were doing with any hierarchy involving inheritance :)

> @CyrusNajmabadi that keeps adding complexity

Depends on perspective i guess.  I find it less complex.  But it's not critical for me.  I have no issue with the approach bein suggested here.  I would like to have analyzers to help catch issues though! :)